rustls/src/msgs/message.rs

```diff
@@ -180,10 +176,66 @@ impl OpaqueMessage {
     const MAX_PAYLOAD: u16 = 16_384 + 2048;
 
     /// Content type, version and size.
-    const HEADER_SIZE: u16 = 1 + 2 + 2;
+    const HEADER_SIZE: usize = 1 + 2 + 2;
 
     /// Maximum on-the-wire message size.
-    pub const MAX_WIRE_SIZE: usize = (Self::MAX_PAYLOAD + Self::HEADER_SIZE) as usize;
+    pub const MAX_WIRE_SIZE: usize = Self::MAX_PAYLOAD as usize + Self::HEADER_SIZE;
+}
+
+#[derive(Clone, Debug)]
+pub struct PrefixedPayload(Vec<u8>);
+
+impl PrefixedPayload {
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut prefixed_payload = Vec::with_capacity(OpaqueMessage::HEADER_SIZE + capacity);
+        prefixed_payload.resize(OpaqueMessage::HEADER_SIZE, 0);
+        Self(prefixed_payload)
+    }
+}
+
+impl AsRef<[u8]> for PrefixedPayload {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[OpaqueMessage::HEADER_SIZE..]
+    }
+}
+
+impl AsMut<[u8]> for PrefixedPayload {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0[OpaqueMessage::HEADER_SIZE..]
+    }
+}
+
+impl<'a> Extend<&'a u8> for PrefixedPayload {
+    fn extend<T: IntoIterator<Item = &'a u8>>(&mut self, iter: T) {
+        self.0.extend(iter)
+    }
+}
+
+impl Deref for PrefixedPayload {
+    type Target = Vec<u8>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for PrefixedPayload {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl From<&[u8]> for PrefixedPayload {
+    fn from(content: &[u8]) -> Self {
+        Self([&[0u8; OpaqueMessage::HEADER_SIZE], content].concat())
```

Comment:
Is `concat()` optimal? I'd probably write this with an explicit `Vec::with_capacity()`.